clients/pkg/promtail/targets/docker/targetmanager.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/grafana/loki/clients/pkg/promtail/targets/target"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/discovery"
+
+	"github.com/grafana/loki/pkg/util"
+)
+
+const (
+	// See github.com/prometheus/prometheus/discovery/moby
+	dockerLabel                = model.MetaLabelPrefix + "docker_"
+	dockerLabelContainerPrefix = dockerLabel + "container_"
+	dockerLabelContainerID     = dockerLabelContainerPrefix + "id"
+	dockerLabelLogStream       = dockerLabelContainerPrefix + "_log_stream"
+)
+
+type TargetManager struct {
+	metrics    *Metrics
+	logger     log.Logger
+	positions  positions.Positions
+	cancel     context.CancelFunc
+	manager    *discovery.Manager
+	pushClient api.EntryHandler
+	groups     map[string]*targetGroup
+}
+
+func NewTargetManager(
+	metrics *Metrics,
+	logger log.Logger,
+	positions positions.Positions,
+	pushClient api.EntryHandler,
+	scrapeConfigs []scrapeconfig.Config,
+) (*TargetManager, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	tm := &TargetManager{
+		metrics:    metrics,
+		logger:     logger,
+		cancel:     cancel,
+		positions:  positions,
+		manager:    discovery.NewManager(ctx, log.With(logger, "component", "docker_discovery")),
+		pushClient: pushClient,
+		groups:     make(map[string]*targetGroup),
+	}
+	configs := map[string]discovery.Configs{}
+	for _, cfg := range scrapeConfigs {
+		if cfg.DockerConfig != nil {
+
+			pipeline, err := stages.NewPipeline(log.With(logger, "component", "docker_pipeline"), cfg.PipelineStages, &cfg.JobName, metrics.reg)
+			if err != nil {
+				return nil, err
+			}
+
+			s := &targetGroup{
+				metrics:       metrics,
+				logger:        logger,
+				targets:       make(map[string]*Target),
+				entryHandler:  pipeline.Wrap(pushClient),
+				defaultLabels: cfg.DockerConfig.Labels,
+				host:          cfg.DockerConfig.Host,
+			}
+			err = s.addTarget(cfg.DockerConfig.ContainerName, model.LabelSet{})
+			if err != nil {
+				return nil, err
+			}
+			tm.groups[cfg.JobName] = s
+		} else if cfg.DockerSDConfigs != nil {
+			pipeline, err := stages.NewPipeline(
+				log.With(logger, "component", "docker_pipeline"),
+				cfg.PipelineStages,
+				&cfg.JobName,
+				metrics.reg,
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			for _, sdConfig := range cfg.DockerSDConfigs {
+				syncerKey := fmt.Sprintf("%s/%s:%d", cfg.JobName, sdConfig.Host, sdConfig.Port)
+				_, ok := tm.groups[syncerKey]
+				if !ok {
+					tm.groups[syncerKey] = &targetGroup{
+						metrics:       metrics,
+						logger:        logger,
+						positions:     positions,
+						targets:       make(map[string]*Target),
+						entryHandler:  pipeline.Wrap(pushClient),
+						defaultLabels: model.LabelSet{},
+						host:          sdConfig.Host,
+					}
+				}
+				configs[syncerKey] = append(configs[syncerKey], sdConfig)
+			}
+		} else {
+			level.Debug(tm.logger).Log("msg", "Docker service discovery configs are empty")
+		}
+	}
+
+	go tm.run()
```

Comment:
Q: how these two goroutines will be stopped?

Reply:
`run` stops when `tm.manager.SyncCh()` is closed. The manager stops when the context is canceled.